File: pset6/credit/credit.py
from cs50 import get_string

def print_result(obj):
    print(obj["comp"])

def check_valid(result, obj):
    if result % 10 == 0:
        obj["valid"] = True
    if obj["valid"] and obj["digits"][0] == "4":
        obj["comp"] = "VISA"
        obj["valid"] = "VALID"
    elif obj["valid"] and obj["digits"][:2] == "34" or obj["digits"][:2] == "37":
        obj["comp"] = "AMEX"
        obj["valid"] = "VALID"
    elif obj["valid"] and obj["digits"][0] == "5":
        obj["comp"] = "MASTERCARD"
        obj["valid"] = "VALID"
    else:
        obj["comp"] = "INVALID"

    print_result(obj)


def calculate_eles(temp, last, obj):
    second_to_last = []

    # Calculate the temporary array.
    for i in temp:
        if len(i) > 1:
            share = int(int(i) / pow(10, len(i) - 1))
            remainder = int(i) % pow(10, len(i) - 1)
            second_to_last.append(share)
            second_to_last.append(remainder)
        else:
            second_to_last.append(int(i))
    # Here is the result of Luhn's Algorithm.
    result = sum(second_to_last) + sum(last)

    check_valid(result, obj)


def reverse_search(number, obj):
    # Make card number to list.
    digits = list(number)
    temp = []
    last = []

    # Process each elements of the array reversely.
    for idx, digit in enumerate(digits[::-1]):
        # If index is second-to-last:
        if idx % 2 == 1:
            temp.append(str(int(digit) * 2))
        else:
            last.append(int(digit))

    calculate_eles(temp, last, obj)


def main():
    card_number = get_string("Number: ")

    # Make a new card dictionary.
    card = {
        "digits": card_number,
        "valid": False,
        "comp": ""
    }

    reverse_search(card_number, card)


main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>


int main(void)
{
    long long bigInt = get_long_long("Number: ");
    int digits = 16;
    int digit, secondToLast, otherDigits;

    otherDigits = 0;
    secondToLast = 0;

    //Get each digits from user's credit card.
    for (int i = 1; i <= digits; i++)
    {
        // Divide credit card number with 10^i and get remainders(digits).
        long long result = powl(10, i);

        digit = (bigInt % result) / pow(10, i - 1);


        // Get second-to-last digits.
        if (i % 2 == 0)
        {
            //Get Products' digits.
            secondToLast += ((digit * 2) / 10) + ((digit * 2) % 10);
        }
        //Get the other digits.
        else
        {
            otherDigits += digit;

        }

    }

    int checkSum = secondToLast + otherDigits;
    int isValid = checkSum % 10;
    int isSixteen = bigInt / powl(10, 14);
    int isFifteen = bigInt / powl(10, 13);
    int isThirteen = bigInt / powl(10, 11);

    //16-digit VISA: first digit is '4' and valid
    if ((isSixteen == 40 || isSixteen == 41 || isSixteen == 42) && isValid == 0)
    {
        printf("VISA\n");
    }

    //13-digit VISA: first digit is '4' and valid
    else if ((isThirteen == 40 || isThirteen == 41 || isThirteen == 42) && isValid == 0)
    {
        printf("VISA\n");
    }


    //15-digit AMEX: first digit is '3' and valid
    else if (isFifteen == 37 && isValid == 0)
    {
        printf("AMEX\n");
    }


    //16-digit MasterCard: first digit is '2' and valid
    else if ((isSixteen == 22 || isSixteen == 51 || isSixteen == 55) && isValid == 0)
    {
        printf("MASTERCARD\n");
    }


    //Unvalid card number.
    else
    {
        printf("INVALID\n");
    }
}

